presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -423,89 +362,164 @@ private static void writePolygon(Geometry geometry, SliceOutput output, boolean
+            output.writeDouble(envelope.getMinX());
+            output.writeDouble(envelope.getMinY());
+            output.writeDouble(envelope.getMaxX());
+            output.writeDouble(envelope.getMaxY());
+        }
     }
 
-    private static void writeEnvelope(Envelope envelope, SliceOutput output)
+    private static void canonicalizePolygonCoordinates(Coordinate[] coordinates, int[] partIndexes, boolean[] shellPart)
     {
-        requireNonNull(envelope, "envelope is null");
-        requireNonNull(output, "output is null");
+        for (int part = 0; part < partIndexes.length - 1; part++) {
+            canonicalizePolygonCoordinates(coordinates, partIndexes[part], partIndexes[part + 1], shellPart[part]);
+        }
+        if (partIndexes.length > 0) {
+            canonicalizePolygonCoordinates(coordinates, partIndexes[partIndexes.length - 1], coordinates.length, shellPart[partIndexes.length - 1]);
+        }
+    }
+
+    private static void canonicalizePolygonCoordinates(Coordinate[] coordinates, int start, int end, boolean isShell)
+    {
+        boolean isClockwise = isClockwise(coordinates, start, end);
 
-        output.writeDouble(envelope.getMinX());
-        output.writeDouble(envelope.getMinY());
-        output.writeDouble(envelope.getMaxX());
-        output.writeDouble(envelope.getMaxY());
+        if ((isShell && !isClockwise) || (!isShell && isClockwise)) {
+            // shell has to be counter clockwise
+            reverse(coordinates, start, end);
+        }
     }
 
     private static boolean isClockwise(Coordinate[] coordinates)
+    {
+        return isClockwise(coordinates, 0, coordinates.length);
+    }
+
+    private static boolean isClockwise(Coordinate[] coordinates, int start, int end)
     {
         // Sum over the edges: (x2 − x1) * (y2 + y1).
         // If the result is positive the curve is clockwise,
         // if it's negative the curve is counter-clockwise.
         double area = 0;
-        for (int i = 1; i < coordinates.length; i++) {
+        for (int i = start + 1; i < end; i++) {
             area += (coordinates[i].x - coordinates[i - 1].x) * (coordinates[i].y + coordinates[i - 1].y);
         }
-        area += (coordinates[0].x - coordinates[coordinates.length - 1].x) * (coordinates[0].y + coordinates[coordinates.length - 1].y);
+        area += (coordinates[start].x - coordinates[end - 1].x) * (coordinates[start].y + coordinates[end - 1].y);
         return area > 0;
     }
+
+    private static void reverse(Coordinate[] coordinates, int start, int end)
+    {
+        for (int i = start; i < start + ((end - start) / 2); i++) {
+            Coordinate buffer = coordinates[i];
+            coordinates[i] = coordinates[start + end - i - 1];
+            coordinates[start + end - i - 1] = buffer;
+        }
+    }
+
+    private enum EsriShapeType
+    {
+        POINT(1),
+        POLYLINE(3),
+        POLYGON(5),
+        MULTI_POINT(8);
+
+        final int code;
```

Comment:
mark as `private`.

Reply:
Outside the world encapsulation is ensured by a class modified. It is `private`. I made this field `public` to emphasize that this field is being directly accessed by the outer class. Unfortunately it is has no effect on compiler level.